### DISCUSSION TITLE
mp_sched_schedule_node vs mp_sched_schedule

### DISCUSSION THREAD

**Discussion by IhorNehrutsa** [micropython/micropython#9855] [Core Development] [answered]
What is the difference in use?
When should one or the other be used?

mp_sched_schedule_node looks like it prevents reentering to an IRQ user function and looks safer for the end user.
But it is used 4 times when the mp_sched_schedule is used 20 times in micropython codebase.

May anybody add some comments to https://github.com/micropython/micropython/blob/master/py/scheduler.c?
Thanks.

P.S. This question applies to ESP32: Add Quadrature Encoder and Pulse Counter classes. #8766

**Comment by jimmo** [comment, marked as answer]
In summary - `mp_sched_schedule` is for running a Python function in the scheduler. The most common use case is to execute a soft IRQ handler. `mp_sched_schedule_node` is for running a C function, typically some sort of background task (bluetooth, networking, usb, etc).

Importantly there are only a fixed number of pre-allocated slots for `mp_sched_schedule`, whereas with `mp_sched_schedule_node` the calling code provides the linked list node.

> mp_sched_schedule_node looks like it prevents reentering to an IRQ user function and looks safer for the end user.

It's really just about the use case -- before `mp_sched_schedule_node`, background tasks implemented in C had to create a "fake" Python function object just so the scheduler could invoke it. `mp_sched_schedule_node` has support for "this thing is already scheduled" (which I think is what you're referring to).

> But it is used 4 times when the mp_sched_schedule is used 20 times in micropython codebase.

mp_sched_schedule_node is newer. There might be some cases of mp_sched_schedule that could now be mp_sched_schedule_node, but for the most part mp_sched_schedule will be used whenever C code needs to invoke a user callback provided in Python.
